cpp/src/reductions/reductions.cpp

```diff
@@ -17,70 +17,91 @@
-  if (col.size() <= col.null_count()) return result;
+  reduce_dispatch_functor(column_view const &col, data_type output_dtype,
+                          rmm::mr::device_memory_resource *mr,
+                          cudaStream_t stream)
+      : col(col), output_dtype(output_dtype), mr(mr), stream(stream) {}
 
-  switch (op) {
-    case reduction_op::SUM:
-      result = reduction::sum(col, output_dtype, mr, stream);
+  template <aggregation::Kind k>
+  std::unique_ptr<scalar> operator()(std::unique_ptr<aggregation> const &agg) {
+    switch (k) {
+    case aggregation::SUM:
+      return reduction::sum(col, output_dtype, mr, stream);
       break;
-    case reduction_op::MIN:
-      result = reduction::min(col, output_dtype, mr, stream);
+    case aggregation::MIN:
+      return reduction::min(col, output_dtype, mr, stream);
       break;
-    case reduction_op::MAX:
-      result = reduction::max(col, output_dtype, mr, stream);
+    case aggregation::MAX:
+      return reduction::max(col, output_dtype, mr, stream);
       break;
-    case reduction_op::ANY:
-      result = reduction::any(col, output_dtype, mr, stream);
+    case aggregation::ANY:
+      return reduction::any(col, output_dtype, mr, stream);
       break;
-    case reduction_op::ALL:
-      result = reduction::all(col, output_dtype, mr, stream);
+    case aggregation::ALL:
+      return reduction::all(col, output_dtype, mr, stream);
       break;
-    case reduction_op::PRODUCT:
-      result = reduction::product(col, output_dtype, mr, stream);
+    case aggregation::PRODUCT:
+      return reduction::product(col, output_dtype, mr, stream);
       break;
-    case reduction_op::SUMOFSQUARES:
-      result =
-          reduction::sum_of_squares(col, output_dtype, mr, stream);
+    case aggregation::SUM_OF_SQUARES:
+      return reduction::sum_of_squares(col, output_dtype, mr, stream);
       break;
-
-    case reduction_op::MEAN:
-      result = reduction::mean(col, output_dtype, mr, stream);
+    case aggregation::MEAN:
+      return reduction::mean(col, output_dtype, mr, stream);
       break;
-    case reduction_op::VAR:
-      result = reduction::variance(col, output_dtype, ddof, mr, stream);
+    case aggregation::VARIANCE: {
+      auto var_agg = static_cast<std_var_aggregation const*>(agg.get());
+      return reduction::variance(col, output_dtype, var_agg->_ddof, mr, stream);
+      }
       break;
-    case reduction_op::STD:
-      result = reduction::standard_deviation(col, output_dtype, ddof, mr, stream);
+    case aggregation::STD: {
+      auto var_agg = static_cast<std_var_aggregation const*>(agg.get());
+      return reduction::standard_deviation(col, output_dtype, var_agg->_ddof, mr, stream);
+      }
       break;
     default:
       CUDF_FAIL("Unsupported reduction operator");
+    }
   }
+};
+
+std::unique_ptr<scalar> reduce(
+    column_view const& col, 
+    std::unique_ptr<aggregation> const &agg,
+    data_type output_dtype,
```

Comment:
Will it be useful if we can derive the output type using the `target_type_t` rather than passing it from API?

Reply:
Seconded. It will probably greatly reduce compile time

Reply:
Hold on here. For series level reductions we specifically wanted the option to be able to specify the desired return type.

Reply:
That's right. In python pandas, dtype can be mentioned as kwargs. 
Last commit is reverted.

Reply:
This is essential for avoiding overflow when summing small types, e.g. 100M int8s...

Reply:
@harrism , actually, `target_type_t` does take care of this by using `int64_t` for all integer reductions.

Reply:
@devavret For all int, it is always `int64_t`. To give a counter example: For `int64_t` column and with python kwarg `dtype=double`, overflow won't happen in this case. if `target_type_t` is used, overflow could happen.

Reply:
@karthikeyann , just clarifying: I backtrack on my original request to use `target_type_t`. I just mentioned this as an fyi for @harrism.